### ISSUE TITLE
Doesn't work under Wayland

### ISSUE THREAD

**Issue by michaldybczak** [smplayer-dev/smplayer#668]
**Describe the bug**
When I open a video file with SMPlayer, it ALWAYS crashes. No matter what format or file, it just refuses to play ANYTHING.

I'm on Manjaro testing, so close to Arch stable repo version 22.7.0-2.
MPV 1:0.35.1-2

System info:
Operating System: Manjaro Linux 
KDE Plasma Version: 5.27.2
KDE Frameworks Version: 5.103.0
Qt Version: 5.15.8
Kernel Version: 6.2.1-2-MANJARO (64-bit)
Graphics Platform: Wayland
Processors: 8 × Intel® Core™ i7-6700HQ CPU @ 2.60GHz
Memory: 7.6 GiB of RAM
Graphics Processor: Mesa Intel® HD Graphics 530
Manufacturer: Alienware
Product Name: Alienware 17 R3
System Version: 1.13.1

I haven't tested X11 session, so this may be a Wayland bug.

I noticed the issue after Plasma 5.27 update but it could have been earlier, because I don't use SMPlayer for every file time. However, Plasma update rested my mime settings and SMPlayer became a default for all video formats and that is why I noticed the issue right away after this update, but it could have been weeks undetected.

Here is the info about some exemplary crash:

```
/usr/bin/mplayer -noquiet -slave -identify -nofs -sub-fuzziness 1 -vo xv,x11, -nodr -double -nomouseinput -input nodefault-bindings:conf=/dev/null -nokeepaspect -wid 5 -monitorpixelaspect 1 -osdlevel 0 -subfont-osd-scale 3 -progbar-align 80 -ass -embeddedfonts -ass-line-spacing 0 -ass-font-scale 1 -noflip-hebrew -ass-styles /home/michaldybczak/.config/smplayer/styles.ass -subcp enca:pl:UTF-8 -subpos 100 -volume 35 -cache 2048 -vf-add screenshot=/home/michaldybczak/Dropbox/Zrzuty ekranu/shot -noslices -af-add volnorm=1 -af-add scaletempo -softvol -softvol-max 160 /home/michaldybczak/Wideo/Blood.and.Treasure.S02E01.720p.WEB.h264-KOGi[eztv.re].mkv

MPlayer SVN-r38407 (C) 2000-2022 MPlayer Team
226 audio & 469 video codecs
do_connect: could not connect to socket
connect: No such file or directory
Failed to open LIRC support. You will not be able to use your remote control.
Terminal type `unknown' is not defined.

Playing /home/michaldybczak/Wideo/Blood.and.Treasure.S02E01.720p.WEB.h264-KOGi[eztv.re].mkv.

Cache fill:  0.00% (0 bytes)   

libavformat version 59.27.100 (external)
libavformat file format detected.
ID_VIDEO_ID=0
[lavf] stream 0: video (h264), -vid 0
ID_AUDIO_ID=0
ID_AID_0_LANG=eng
[lavf] stream 1: audio (eac3), -aid 0, -alang eng
ID_SUBTITLE_ID=0
ID_SID_0_LANG=eng
ID_SID_0_NAME=English (SDH)
[lavf] stream 2: subtitle (srt), -sid 0, -slang eng, English (SDH)
VIDEO:  [H264]  1280x720  0bpp  24.000 fps    0.0 kbps ( 0.0 kbyte/s)
[ass] Error opening iconv descriptor
X11 error: BadMatch (invalid parameter attributes)
Opening video filter: [ass auto=1]
[ass] auto-open
Opening video filter: [screenshot=/home/michaldybczak/Dropbox/Zrzuty ekranu/shot]
[ass] libass API version: 0x1701000
[ass] libass source: commit: 0.17.1-0-ge8ad72accd3a84268275a9385beb701c9284e5b3-dirty
[ass] Shaper: FriBidi 1.0.12 (SIMPLE) HarfBuzz-ng 7.0.1 (COMPLEX)
[ass] Using font provider fontconfig
==========================================================================
Opening video decoder: [ffmpeg] FFmpeg's libavcodec codec family
libavcodec version 59.37.100 (external)
Selected video codec: [ffh264] vfm: ffmpeg (FFmpeg H.264)
==========================================================================
ID_VIDEO_CODEC=ffh264
Clip info:
 encoder: libebml v1.4.2 + libmatroska v1.6.4
ID_CLIP_INFO_NAME0=encoder
ID_CLIP_INFO_VALUE0=libebml v1.4.2 + libmatroska v1.6.4
ID_CLIP_INFO_N=1
Load subtitles in /home/michaldybczak/Wideo/
ID_FILENAME=/home/michaldybczak/Wideo/Blood.and.Treasure.S02E01.720p.WEB.h264-KOGi[eztv.re].mkv
ID_DEMUXER=lavfpref
ID_VIDEO_FORMAT=H264
ID_VIDEO_BITRATE=0
ID_VIDEO_WIDTH=1280
ID_VIDEO_HEIGHT=720
ID_VIDEO_FPS=24.000
ID_VIDEO_ASPECT=0.0000
ID_AUDIO_FORMAT=EAC3
ID_AUDIO_BITRATE=640000
ID_AUDIO_RATE=48000
ID_AUDIO_NCH=6
ID_START_TIME=0.00
ID_LENGTH=2552.88
ID_SEEKABLE=1
ID_CHAPTERS=0
==========================================================================
Opening audio decoder: [ffmpeg] FFmpeg/libavcodec audio decoders
AUDIO: 48000 Hz, 2 ch, floatle, 640.0 kbit/20.83% (ratio: 80000->384000)
ID_AUDIO_BITRATE=640000
ID_AUDIO_RATE=48000
ID_AUDIO_NCH=2
Selected audio codec: [ffeac3] afm: ffmpeg (FFmpeg E-AC-3)
==========================================================================
[AO OSS] audio_setup: Can't open audio device /dev/dsp: No such file or directory
AO: [alsa] 48000Hz 2ch floatle (4 bytes per sample)
ID_AUDIO_CODEC=ffeac3
[Mixer] No hardware mixing, inserting volume filter.
Starting playback...
Movie-Aspect is 1.78:1 - prescaling to correct movie aspect.
ID_VIDEO_ASPECT=1.7778
[swscaler @ 0x7f4bb4faab00]bicubic scaler, from yuv420p to rgb24 using MMXEXT
[swscaler @ 0x7f4bb4faab00]using unscaled yuv420p -> rgb24 special converter
VO: [xv] 1280x720 => 1280x720 Planar YV12 
X11 error: BadWindow (invalid Window parameter)
X11 error: BadWindow (invalid Window parameter)
X11 error: BadDrawable (invalid Pixmap or Window parameter)
X11 error: BadWindow (invalid Window parameter)
X11 error: BadDrawable (invalid Pixmap or Window parameter)
X11 error: BadWindow (invalid Window parameter)
X11 error: BadDrawable (invalid Pixmap or Window parameter)
X11 error: BadWindow (invalid Window parameter)

X11 error: BadDrawable (invalid Pixmap or Window parameter)
X11 error: BadDrawable (invalid Pixmap or Window parameter)
X11 error: BadWindow (invalid Window parameter)


MPlayer interrupted by signal 8 in module: filter video
ID_SIGNAL=8
- MPlayer crashed by bad usage of CPU/FPU/RAM.
  Recompile MPlayer with --enable-debug and make a 'gdb' backtrace and
  disassembly. Details in DOCS/HTML/en/bugreports_what.html#bugreports_crash.
- MPlayer crashed. This shouldn't happen.
  It can be a bug in the MPlayer code _or_ in your drivers _or_ in your
  gcc version. If you think it's MPlayer's fault, please read
  DOCS/HTML/en/bugreports.html and follow the instructions there. We can't and
  won't help unless you provide this information when reporting a possible bug.
```

Here is info from the terminal translated loosely on English:

This is SMPlayer wer. 22.7.0 (version 10091) launched on Linux
kf.coreaddons: "Could not load plugin from /usr/lib/qt/plugins/ffmpegthumbs.so: Incompatibility of metadata verification in plugin \"/usr/lib/qt/plugins/ffmpegthumbs.so\""

Hmm... It may be an upstream issue with ffmpeg 2:5.1.2-12

**Comment by smplayer-dev** [comment]
Maybe it's a bug in mplayer. Does the video play if open it directly with mplayer?
Does it work if you change the multimedia engine (preferences -> general) to mpv?

**Comment by michaldybczak** [comment]
MPV Player works correctly for the same files.

When I set engine to MPV, I got a black screen and heard only audio. When I forwarded the video, SMPlayer froze and in the end showed the same crash error message.

By the way, I noticed, that the engine setting doesn't stick. I see "other" and when I set it to MPV or MPlayer, restart the app, it still says "other".

Is this bug not reported by anyone from Manjaro or Arch? I'm not a member of Arch forums, so I can't check. I don't see any recent SMPlayer topic on Manjaro forums either. Usually, when there is an issue with SMPlayer, people start to write about it.

It can be a versioning issue on Manjaro. Meaning, SMPlayer wasn't updated since 21st November 2022, while MPV was updated in 22nd February this year. Maybe SMPlayer needs to recompiled? Let me check. I will install git version.

Nope, git version also doesn't work, although it shows a different error message. It behaves a little different too. Instead of giving error right after start, it plays sound with black screen, then freezes. When I click on X to close the app, it then opens error info after a while. Here it is:

```
/usr/bin/mpv --no-quiet --terminal --no-msg-color --input-ipc-server=/tmp/smplayer-mpv-3c31 --msg-level=ffmpeg/demuxer=error --video-rotate=no --no-config --no-fs --hwdec=no --sub-auto=fuzzy --vo=xv,x11, --no-input-default-bindings --input-vo-keyboard=no --no-input-cursor --cursor-autohide=no --no-keepaspect --wid=5 --monitorpixelaspect=1 --osd-level=1 --osd-scale=1 --osd-bar-align-y=0.6 --sub-ass --embeddedfonts --sub-ass-line-spacing=0 --sub-scale=1 --sub-font=Arial --sub-color=#ffffffff --sub-shadow-color=#ff000000 --sub-border-color=#ff000000 --sub-border-size=0.75 --sub-shadow-offset=2.5 --sub-font-size=65 --sub-bold=no --sub-italic=no --sub-margin-y=8 --sub-margin-x=20 --sub-codepage=UTF-8 --sub-pos=100 --volume=56 --demuxer-max-bytes=2097152 --screenshot-template=cap_%F_%p_%02n --screenshot-format=jpg --screenshot-directory=/home/michaldybczak/Dropbox/Zrzuty ekranu --af-add=lavfi=[acompressor] --audio-pitch-correction=yes --volume-max=160 --term-playing-msg=MPV_VERSION=${=mpv-version:}
INFO_VIDEO_WIDTH=${=width}
INFO_VIDEO_HEIGHT=${=height}
INFO_VIDEO_ASPECT=${=video-params/aspect}
INFO_VIDEO_FPS=${=container-fps:${=fps}}
INFO_VIDEO_FORMAT=${=video-format}
INFO_VIDEO_CODEC=${=video-codec}
INFO_DEMUX_ROTATION=${=track-list/0/demux-rotation}
INFO_AUDIO_FORMAT=${=audio-codec-name}
INFO_AUDIO_CODEC=${=audio-codec}
INFO_AUDIO_RATE=${=audio-params/samplerate}
INFO_AUDIO_NCH=${=audio-params/channel-count}
INFO_LENGTH=${=duration:${=length}}
INFO_DEMUXER=${=current-demuxer:${=demuxer}}
INFO_SEEKABLE=${=seekable}
INFO_TITLES=${=disc-titles}
INFO_CHAPTERS=${=chapters}
INFO_TRACKS_COUNT=${=track-list/count}
METADATA_TITLE=${metadata/by-key/title:}
METADATA_ARTIST=${metadata/by-key/artist:}
METADATA_ALBUM=${metadata/by-key/album:}
METADATA_GENRE=${metadata/by-key/genre:}
METADATA_DATE=${metadata/by-key/date:}
METADATA_TRACK=${metadata/by-key/track:}
METADATA_COPYRIGHT=${metadata/by-key/copyright:}
INFO_MEDIA_TITLE=${=media-title:}
INFO_STREAM_PATH=${stream-path}
 --audio-client-name=SMPlayer --term-status-msg=STATUS: ${=time-pos} / ${=duration:${=length:0}} P: ${=pause} B: ${=paused-for-cache} I: ${=core-idle} VB: ${=video-bitrate:0} AB: ${=audio-bitrate:0} /home/michaldybczak/Wideo/Mary.Kills.People.S03E03.720p.HDTV.x264-aAF[eztv].mkv

 (+) Video --vid=1 (*) (h264 1280x720 23.976fps)
 (+) Audio --aid=1 (*) (ac3 6ch 48000Hz)
[vo/xv/x11] X11 error: BadDrawable (invalid Pixmap or Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 5, serial: 12
[vo/xv/x11] Error code: 9, request code: e, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 5, serial: 13
[vo/xv/x11] Error code: 3, request code: 28, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 5, serial: 14
[vo/xv/x11] Error code: 3, request code: 2, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 5, serial: 18
[vo/xv/x11] Error code: 3, request code: 1, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 1b
[vo/xv/x11] Error code: 3, request code: 12, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 1e
[vo/xv/x11] Error code: 3, request code: 92, minor code: 3
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 20
[vo/xv/x11] Error code: 3, request code: 12, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 21
[vo/xv/x11] Error code: 3, request code: 12, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 23
[vo/xv/x11] Error code: 3, request code: 12, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 24
[vo/xv/x11] Error code: 3, request code: 12, minor code: 0
[vo/xv] Warning: this legacy VO has bad quality and performance, and will in particular result in blurry OSD and subtitles. You should fix your graphics drivers, or not force the xv VO.
AO: [pipewire] 48000Hz 5.1(side) 6ch double
VO: [xv] 1280x720 yuv420p
[vo/xv/x11] X11 error: BadDrawable (invalid Pixmap or Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 5, serial: 37
[vo/xv/x11] Error code: 9, request code: e, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 5, serial: 38
[vo/xv/x11] Error code: 3, request code: 28, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 39
[vo/xv/x11] Error code: 3, request code: c, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 3b
[vo/xv/x11] Error code: 3, request code: 12, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 3c
[vo/xv/x11] Error code: 3, request code: 2, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 3d
[vo/xv/x11] Error code: 3, request code: 2, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 3e
[vo/xv/x11] Error code: 3, request code: 3, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 40
[vo/xv/x11] Error code: 3, request code: 8, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 41
[vo/xv/x11] Error code: 3, request code: 12, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 44
[vo/xv/x11] Error code: 3, request code: 3, minor code: 0
[vo/xv/x11] X11 error: BadDrawable (invalid Pixmap or Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 5, serial: 46
[vo/xv/x11] Error code: 9, request code: e, minor code: 0
[vo/xv/x11] X11 error: BadWindow (invalid Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 5, serial: 47
[vo/xv/x11] Error code: 3, request code: 28, minor code: 0
[vo/xv/x11] X11 error: BadDrawable (invalid Pixmap or Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 48
[vo/xv/x11] Error code: 9, request code: 37, minor code: 0
[vo/xv/x11] X11 error: BadDrawable (invalid Pixmap or Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 49
[vo/xv/x11] Error code: 9, request code: 37, minor code: 0
[vo/xv/x11] X11 error: BadDrawable (invalid Pixmap or Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 57
[vo/xv/x11] Error code: 9, request code: 95, minor code: 13
[vo/xv/x11] X11 error: BadDrawable (invalid Pixmap or Window parameter)
[vo/xv/x11] Type: 0, display: 0x7f3758044540, resourceid: 1800002, serial: 58
[vo/xv/x11] Error code: 9, request code: 95, minor code: 13
[vo/xv] X11 can't keep up! Waiting for XShm completion events...
MPV_VERSION=mpv 0.35.1
INFO_VIDEO_WIDTH=1280
INFO_VIDEO_HEIGHT=720
INFO_VIDEO_ASPECT=1.777778
INFO_VIDEO_FPS=23.976025
INFO_VIDEO_FORMAT=h264
INFO_VIDEO_CODEC=h264 (H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10)
INFO_DEMUX_ROTATION=
INFO_AUDIO_FORMAT=ac3
INFO_AUDIO_CODEC=ac3 (ATSC A/52A (AC-3))
INFO_AUDIO_RATE=48000
INFO_AUDIO_NCH=6
INFO_LENGTH=2615.872000
INFO_DEMUXER=mkv
INFO_SEEKABLE=yes
INFO_TITLES=
INFO_CHAPTERS=0
INFO_TRACKS_COUNT=2
METADATA_TITLE=
METADATA_ARTIST=
METADATA_ALBUM=
METADATA_GENRE=
METADATA_DATE=
METADATA_TRACK=
METADATA_COPYRIGHT=
INFO_MEDIA_TITLE=Mary.Kills.People.S03E03.720p.HDTV.x264-aAF[eztv].mkv
INFO_STREAM_PATH=/home/michaldybczak/Wideo/Mary.Kills.People.S03E03.720p.HDTV.x264-aAF[eztv].mkv
INFO_VIDEO_DSIZE=1280x720
INFO_TRACK_0: video 1 '' '' yes
INFO_TRACK_1: audio 1 '' '' yes
[mkv] Too many packets in the demuxer packet queues:
[mkv]   video/0: 194 packets, 2115520 bytes
[mkv]   audio/1: 0 packets, 0 bytes
Exiting... (Quit)
```

Is there a way to recompile it with debug symbols, let me know. Remeber I use AUR to install git version.

**Comment by michaldybczak** [comment]
Ah, I realized the error complains about X11, while it is a Wayland session!

I re-logged to X11 and SMPlayer works fine there. So the issue is with SMPlayer not working under the Wayland or not connecting with X11-Wayland protocol or whatever it is called.

**Comment by smplayer-dev** [comment]
Have you checked if it works on Wayland if you use a flatpak or snap package?
https://blog.smplayer.info/how-to-fix-the-video-problem-on-wayland/

**Comment by michaldybczak** [comment]
Snap version works.
So is it an upstream issue? Will the patch be included in future mpv and mplayer version?
I would prefer to use repo version as snaps launch longer, flatpacks use too much space and both are using file window (not sure how to name it) which isn't integrated with the system and looks like wine app. Those are usable, but if repo version is just better in any way.

EDIT: Setting Video driver to x11 didn't help.

**Comment by smplayer-dev** [comment]
No, they aren't going to add the patch. The patch is actually based on some old code that has already been removed. mpv provides a better solution but in order to be used by smplayer it will probably require significant changes, which I have not yet had time to implement. Therefore, consider this implementation a hack rather than a proper one.

**Comment by michaldybczak** [comment]
OK, so if that is a partial upstream issue and the fix in SMPlayer won't come in the near future, I'm closing this case. 

However, more and more people are switching to Wayland, so this will become a bigger problem with time. Luckily, SMPlayer is oftentimes used as a fallback if VLC has some issues (which happens from time to time), so it's not a major issue, but it's a pity that it's not resolved properly.

**Comment by smplayer-dev** [comment]
I understand your point. However, I'm not sure about what it's the best solution. Let's see. The proper solution would be to use the libmpv but this would require significant changes to smplayer. Also, even if we implement this solution, it won't fix the problem for mplayer since it doesn't have a similar library.

On Windows and Mac OS, the smplayer packages include builds of mpv and mplayer, along with the required libraries. I can provide versions that I know work well with smplayer. However, on Linux, there are many different distros with many versions, and I often receive bug reports for smplayer that are actually bugs in the packages of mpv/mplayer or their dependencies. For example, there's a recent bug about the floating control not being visible in Ubuntu 22.10, which is likely caused by a bug in a system library.

Using a snap or flatpak package would allow me to provide a more controlled environment that is less susceptible to bugs in other regular packages.
However, I am aware that snap and flatpak packages can have their own issues,  like not good integration with the desktop, but maybe that could be sorted out.

**Comment by michaldybczak** [comment]
Thanks for the info.
I'm not a developer, so I won't understand all the details and dependencies. However, I understand that the biggest issue is that you can't maintain to the level it would work well everywhere. That is the reality for many FOSS projects. We may not like it, but it is what it is.
Debugging is also complicated, because it's often very tricky to decide what is at fault. It could be an error in the library or a conscious feature of that library, and the app should have a patch that would make it work nevertheless. I'm just theorizing.

SMPlayer is a valued app, because it works differently than VLC, and because of that it tends to work when VLC isn't.

As to theming, SMPlayer is not good at that, so there is no issue here... ironically. I mean, the default skin is just horrible and outdated, but one can test various skins and choose one that is more or less agreeable with system look. That way, I can have the same look on repo package and snap. I personally use Mini GUI Breeze Fusion. Minimalistic and modern. That said, snap or flatpacks shouldn't give any theming issues, because SMPlayer is not an app that would integrate with system theming anyway.

I don't think that all Linux distro will make snaps/flatpacks default. There will be bastions of regular compiled packages. Just as Gentoo keeps it like it was originally, Arch and other rolling releases, will continue having own repo packages. LTS distros are a different thing. However, some packages will remain available only as snap/flatpack if the packaged version stop working properly.

Anyway, let's leave it as it is for now. Maybe some solution will come with time.
Thanks again.

**Comment by jayaddison** [comment]
Hi @michaldybczak @smplayer-dev: I don't know whether this information will be helpful for you, but I encountered this same issue today, and have found a few clues about the cause (it appears to be a bug in `mplayer`, but it is possible that `smplayer` could add a workaround in the "Wayland workarounds" handling in the code).

In particular, it seems to be the combination of two `mplayer` options:

- `wid`, to attach `mplayer` to an existing window (unsupported during Wayland sessions, I think - explaining the `X11 error: BadWindow` messages)
- `nokeepaspect` - allowing arbitrary resizing of the `mplayer` video window

The error (`Thread 1 "mplayer" received signal SIGFPE, Arithmetic exception.
`) is a floating-point exception that occurs here in the `draw_osd` function -- when attempting to draw some on-screen-display text:

https://sources.debian.org/src/mplayer/2%3A1.5%2Bsvn38408-1/libvo/vo_xv.c/#L365-L367

My theory is that because `wid` attempts to attach to a window, and expects to find a width and height for that window -- but fails, some dimension values are left uninitialized/zero.

If `nokeepaspect` is removed from the command-line (using the default for `mplayer`: keep the video's aspect ratio), then those variables appear to have different values.

In that case, the video will still not be displayed (because `mplayer` does not find a window to attach to), but it will not crash, and audio does play.

If that is true then a workaround could be not to set the `wid` value during Wayland sessions.

**Comment by michaldybczak** [comment]
@smplayer-dev, should I reopen the issue? Does the previous post shows promise to create a workaround?

However, if the issue is with mplayer, should we submit a bug to mplayer?

**Comment by jayaddison** [comment]
@michaldybczak I think I should collect some more detailed information and open that as a bug with `mplayer` - I've found maybe 80% of the information for that, and with another hour or two of work (hopefully within the next week or two), I think I can provide a detailed bugreport and maybe a patch to fix the issue there.

Whether to re-open this.. I would agree that is up to @smplayer-dev.  It can take time to confirm and develop workarounds like this (not to mention maintain them over a potentially long duration).

**Comment by smplayer-dev** [comment]
I workaround that doesn't use -wid has been already implemented in the flatpak, snap and appimage packages you can find here:
https://github.com/smplayer-dev/smplayer/releases

**Comment by michaldybczak** [comment]
Can't it be applied to a normal package? Sorry if that is a dumb question. I'm not a developer.

**Comment by smplayer-dev** [comment]
It's possible, the problem is that it would require to install a patched mplayer, which may be overwritten by the system during updates. The flatpak, snap and appimage packages allow me to provide that patched mplayer which won't be overwritten by system updates.

**Comment by michaldybczak** [comment]
So if the issue is upstream (mplayer), why won't we address the issue there? Is it maintained? I mean, the mplayer.

**Comment by smplayer-dev** [comment]
Or rather the problem is on Wayland, because on X11 `-wid` works well.

**Comment by jayaddison** [comment]
Hi - I've reported the crash-under-Wayland bug to Debian (because it affects Debian's packaged `bookworm` version), and a maintainer there has filed an `mplayer` ticket about the `wid` / `nokeepaspect` problem that causes it:

- https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1034867
- https://trac.mplayerhq.hu/ticket/2411

**Comment by jayaddison** [comment]
Note: fixing those bugs will prevent the program from crashing due to the floating point exception, but they won't enable Wayland compatibility.  I don't yet know what the best way to achieve that will be.

**Comment by jayaddison** [comment]
The larger compatibility issue with (X)Wayland, I think, is due to the fact that the `wid` parameter is retrieved by calling a [Qt `winId()` function](https://github.com/smplayer-dev/smplayer/blob/v22.7.0/src/core.cpp#L1863) that doesn't correspond to an X11 window identifier.

That could explain the `X11 error: BadWindow (invalid Window parameter)` error messages.

When I run `smplayer` and try to play a video using the `mplayer` output driver, the `wid` parameter is often a low integer value (like `3`, `4`, `5`, ...), but we would probably generally expect a larger, hex integer representation (like `0x6000ab` for example).

I think that that is a known limitation / problem with Qt, and https://bugreports.qt.io/browse/QTBUG-58560 seems to be a relevant bugreport (note that it mentions that the returned window identifiers are in fact a local integer counter.. that sounds like it would explain the low-valued-integer `wid` values).

**Comment by rdoeffinger** [comment]
Largely, wayland just does not support this operating model with multiple processes.
Have you tried QT_QPA_PLATFORM=xcb environment variable to make just SMPlayer use X11 even when running on Wayland?
Might even be possible for SMPlayer to programmatically detect that it is running on wayland and force X11 as Qt platform?
(see also https://trac.mplayerhq.hu/ticket/2411 )

**Comment by fenglengshun** [comment]
Alternatively, since smplayer already ships multiple .desktop files (for enqueue), what's the harm in shipping one more, an "smplayer (X11)" with the appropriate env variable, as a temporary solution?

**Comment by smplayer-dev** [comment]
What distribution has this problem? 
I think  QT_QPA_PLATFORM=xcb is already the default.

**Comment by fenglengshun** [comment]
I just installed smplayer in an arch distrobox, and `/usr/share/applications/smplayer.desktop` still only have `Exec=smplayer %U`

**Comment by smplayer-dev** [comment]
Have you tried the flatpak or the snap packages? Those should work on Wayland.

**Comment by fenglengshun** [comment]
I am using the flatpak packages. Sorry if I was unclear, I simply brought up the idea of exporting a .desktop file that forces x11 as a solution since it seems some people are still having an issue and might not want to use Flatpak or snaps (I have no problem with either).

Edit: I currently run ublue with a Fedora Kinoite 38 base.

**Comment by smplayer-dev** [comment]
Testing on Ubuntu 22.10, with Wayland, if I run smplayer from command line I get this warning:
`Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway`
So it seems that QT_QPA_PLATFORM=xcb is the default, at least on this distro.

**Comment by michaldybczak** [comment]
I experienced this problem on Manjaro, so with Arch packages, which are as vanilla as possible, so I suspect there were no xcb by default. Currently, I have snap installed, so it works, but as snaps go, the launching is slow.

**Comment by rdoeffinger** [comment]
Since the comments left me a bit confused I tried myself on Kubuntu 23.04
Starting smplayer from command-line resulted in the issues described.
QT_QPA_PLATFORM=xcb smplayer
works fine.
That also matches the description in issue #721 .
Just setting that environment variable seems a rather partial fix though, as there are many different ways to start smplayer.
There does not seem to be a way to disable the wayland platform plugin specifically. Could actually not even find a way to select a platform. So maybe the closest is to check if we are on Linux and QT_QPA_PLATFORM is not set, and if so set it to xcb, somewhere before QApplication is created?

**Comment by smplayer-dev** [comment]
Implemented as you suggested in https://github.com/smplayer-dev/smplayer/commit/a4dd1953c0f2ad9bdcb455279d3409d3ce68873b.

**Comment by luiscastro193** [comment]
I still think some proper wayland support with libmpv would be great. Even if it means dropping support for mplayer. After almost 15 years using and loving smplayer, I have to temporally move to native mpv for this issue (due to subtitles rendering and fractional scaling issues on gnome xwayland).

**Comment by fccfc5d245c** [comment]
In case it helps, I'm adding that the Flatpack version, installed on Debian 12 through Discover, worked out of the box for me, with Wayland support checked and the default Output driver. 

And I couldn't play a single video with the .deb version, under x11 or any settings. Thank you, guys.

**Comment by fancywriter** [comment]
I have the same problem on smplayer-22.2.0 with Wayland (in Xorg works fine).
Setting variable QT_QPA_PLATFORM=xcb before starting smplayer indeed helped to make it work.

**Comment by smplayer-dev** [comment]
Check version [23.6.0](https://github.com/smplayer-dev/smplayer/releases/tag/v23.6.0)

**Comment by fancywriter** [comment]
Yes, seems that it was fixed in 23.6.0, thanks!
I hope my comment might be useful for those who cannot easily upgrade by any reason.

**Comment by michaldybczak** [comment]
I confirm. With 23.6.0 it works on Wayland. Problem solved. :)
Thank you all.

**Comment by Francewhoa** [comment]
Same challenge with SMPlayer 23.6.0 on Debian 12 with Wayland. To fix this challenge, using [Flatseal](https://flathub.org/en/apps/com.github.tchx84.Flatseal), I added the **Environment variable** `QT_QPA_PLATFORM=wayland`.
